machineconfiguration/v1alpha1/types_machineosconfig.go

```diff
@@ -0,0 +1,264 @@
+type ImageSecretObjectReference struct {
+	// name is the name of the secret used to push or pull this MachineOSConfig object.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+}
+
+type TriggerRebuild string
+
+const (
+	// RebuildNow indicates that the user wants to rebuild their image immediately
+	RebuildNow TriggerRebuild = "RebuildNow"
+	// RebuildDelayed indicates that the user wants to rebuild their image the next time the build controller is idle.
+	// if there is a build in progress, a rebuild will be triggered the next time that build fails or succeeds. This is often used
+	// when the user changes the top level configuration for a pool's build.
+	RebuildDelayed TriggerRebuild = "RebuildDelayed"
+	// NoRebuild indicates the user does not want a manual rebuild, this is the default.
+	NoRebuild TriggerRebuild = "NoRebuild"
+)
+
+// MachineOSRebuild describes user options to manage rebuilding the image.
+type MachineOSRebuild struct {
+	// rebuildStrategy is the user provided manual rebuild method of choice
+	// +kubebuilder:default:=NoRebuild
+	// +kubebuilder:validation:Enum:=NoRebuild;RebuildNow;RebuildDelayed
+	// +kubebuilder:validation:Required
+	RebuildStrategy TriggerRebuild `json:"rebuildStrategy"`
+	// maxRetries is the amount of times we want to auto-rebuild after build failures before we deem the update unreconcilable.
+	// +kubebuilder:default:=3
+	// +kubebuilder:validation:Required
+	MaxRetries int `json:"maxRetries"`
+}
+
+type MachineOSImageBuilderType string
+
+const (
+	// describes that the machine-os-builder will use the built in OpenshiftImageBuilder
+	OCPBuilder MachineOSImageBuilderType = "OpenShiftImageBuilder"
+	// describes that the machine-os-builder will use a custom pod builder that uses buildah
+	PodBuilder MachineOSImageBuilderType = "PodImageBuilder"
+	// Default means the cluster will use the OpenshiftImageBuilder
+	DefaultBuilder MachineOSImageBuilderType = "Default"
+)
+
+// BuildHistory contains information about related builds
+// these builds can be failed, interrupted, or succeeded.
+type BuildHistory struct {
+	// name is the name of the build
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+	// buildFailure contains an optional message of why this build ended prematurely.
+	// +optional
+	BuildFailure string `json:"buildFailure"`
+	// configMaps references all config map objects created during the build process
+	// +patchMergeKey=name
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=name
+	// +optional
+	ConfigMaps []BuildConfigMapObjectReference `json:"configmaps" patchStrategy:"merge" patchMergeKey:"name" `
+	// buildPod references the build pod used in this previous build if it existed
+	// +optional
+	BuildPod BuildPodReference `json:"buildPod"`
+}
+
+// BuildPodReference refers to the pod used to build an image that either failed or was interrupted
+type BuildPodReference struct {
```

Comment:
this is used for status. I have is in the config file since the MOSC uses it in BuildHistory. Honestly, idk if we need this (or configmaps) @cheesesashimi voted to add those in, but as I am implementing in MCO, seems to be causing friction. Would be ok with removing this AND configmaps in the build history. @hexfusion